### DISCUSSION TITLE
.ply point cloud export

### DISCUSSION THREAD

**Discussion by cpkcpkcpk** [NVlabs/instant-ngp#402] [Q&A]
Hey,

Is there any possibility of adding .ply point cloud export of the NeRF as it is shown in the viewport, without meshing?

Thanks

**Comment by Tom94** [comment]
Hi! You can use the Python bindings to get just the vertices of the mesh (without adjacency data) and then use a Python .ply module to export that. Cheers!

**Comment by bonfry** [comment]
I found this method for getting vertices:`testbed.aabb.get_vertices()` but how can | export all vertices from Bounding box. I tried this:

```python
import numpy as np

args = np.array([testbed.aabb.min,testbed.aabb.max])
testbed.aabb.get_vertices(args)
```
but I receive this error:
```
TypeError: get_vertices(): incompatible function arguments. The following argument types are supported:
    1. (self: pyngp.BoundingBox, arg0: numpy.ndarray[numpy.float32[3, 1]]) -> None

Invoked with: <pyngp.BoundingBox object at 0x7f8ff5af3230>, array([[-1.5, -1.5, -1.5],
       [ 2.5,  2.5,  2.5]], dtype=float32)
```

**Comment by satyajit-ink** [comment]
@bonfry i got it working

`pip install plyfile`

```python
import pyngp as ngp
import numpy as np
from plyfile import PlyData, PlyElement

mode = ngp.TestbedMode.Nerf
testbed = ngp.Testbed(mode)
testbed.load_snapshot("../data/room_dslr/base.msgpack")
mc = testbed.compute_marching_cubes_mesh()

vertex = np.array(list(zip(*mc["V"].T)), dtype=[('x', 'f4'), ('y', 'f4'), ('z', 'f4')])
vertex_color = np.array(list(zip(*((mc["C"] * 255).T))), dtype=[('red', 'u1'), ('green', 'u1'), ('blue', 'u1')])

n = len(vertex)
assert len(vertex_color) == n

vertex_all = np.empty(n, vertex.dtype.descr + vertex_color.dtype.descr)

for prop in vertex.dtype.names:
    vertex_all[prop] = vertex[prop]

for prop in vertex_color.dtype.names:
    vertex_all[prop] = vertex_color[prop]

ply = PlyData([PlyElement.describe(vertex_all, 'vertex')], text=False)

ply.write('nerf_pc.ply')
```

now you can open it with something like `meshlab`

**Reply by qungvinh1280** [reply]
> @bonfrytôi đã làm cho nó hoạt động
> 
> `pip install plyfile`
> 
> ```python
> import pyngp as ngp
> import numpy as np
> from plyfile import PlyData, PlyElement
> 
> mode = ngp.TestbedMode.Nerf
> testbed = ngp.Testbed(mode)
> testbed.load_snapshot("../data/room_dslr/base.msgpack")
> mc = testbed.compute_marching_cubes_mesh()
> 
> vertex = np.array(list(zip(*mc["V"].T)), dtype=[('x', 'f4'), ('y', 'f4'), ('z', 'f4')])
> vertex_color = np.array(list(zip(*((mc["C"] * 255).T))), dtype=[('red', 'u1'), ('green', 'u1'), ('blue', 'u1')])
> 
> n = len(vertex)
> assert len(vertex_color) == n
> 
> vertex_all = np.empty(n, vertex.dtype.descr + vertex_color.dtype.descr)
> 
> for prop in vertex.dtype.names:
>     vertex_all[prop] = vertex[prop]
> 
> for prop in vertex_color.dtype.names:
>     vertex_all[prop] = vertex_color[prop]
> 
> ply = PlyData([PlyElement.describe(vertex_all, 'vertex')], text=False)
> 
> ply.write('nerf_pc.ply')
> ```
> 
> bây giờ bạn có thể mở nó bằng một cái gì đó như`meshlab`

I am trying to run your file but it shows error:
![image](https://user-images.githubusercontent.com/125844116/225806258-4f0f22f3-3c43-481f-a3a0-2c68aacdb18d.png)

**Comment by satyajit-ink** [comment]
Update: 

the resolution can be computed like this

```python
def round_up_to_base(x, base=10):
    return x + (base - x) % base

def get_marching_cubes_res(res_1d: int, aabb:  ngp.BoundingBox ) -> np.ndarray:
	scale = res_1d / (aabb.max - aabb.min).max()
	res3d = (aabb.max - aabb.min) * scale + 0.5
	res3d = round_up_to_base(res3d.astype(np.int32), 16)
	return res3d

mc = testbed.compute_marching_cubes_mesh(resolution=get_marching_cubes_res(512, testbed.aabb), aabb=testbed.aabb, thresh=2)

```

**Reply by bonfry** [reply]
Thanks :D

**Reply by TheShoes** [reply]
I got this working and its great.  
This is first taking the nerf ==> meshing it ==> using that data to create a new point cloud? 
it just does not have the same fidelity like inside the nerf program 
im sure there is some fancy processing they are doing for the camera but still... is it my scaling or some export setting or its just the way it is?

**Reply by TheShoes** [reply]
also how would i make the resolution higher in the point cloud?  what numbers would i change in the above code?

**Reply by satyajit-ink** [reply]
> also how would i make the resolution higher in the point cloud? what numbers would i change in the above code?

```python
mc = testbed.compute_marching_cubes_mesh(resolution=get_marching_cubes_res(512, testbed.aabb), aabb=testbed.aabb, thresh=2)
```

here change the `512` to whatever you like.

**Reply by Daphoeno** [reply]
I tried playing around with this code to compmute the marching cubes mesh for multiple different chunks of the NeRF data by changing the aabb bounds value and whilst it seemed to crop the mesh correctly the result is a bit messed up both with the point positions/resolution and the vertex colours

do you have any suggestion on how I could export the point cloud for a given region in the testbed bounds by changing the above code?

**Reply by Mehi44** [reply]
Thanks for great solution ! it worked for me too, but when I try to change the resolution I get this : 

```
import pyngp as ngp
import numpy as np
from plyfile import PlyData, PlyElement

mode = ngp.TestbedMode.Nerf
testbed = ngp.Testbed(mode)
testbed.load_snapshot("/home/mehi/Nerfs/instant-ngp/saved/office.msgpack")

def round_up_to_base(x, base=10):
    return x + (base - x) % base

def get_marching_cubes_res(res_1d: int, aabb:  ngp.BoundingBox ) -> np.ndarray:
	scale = res_1d / (aabb.max - aabb.min).max()
	res3d = (aabb.max - aabb.min) * scale + 0.5
	res3d = round_up_to_base(res3d.astype(np.int32), 16)
	return res3d

mc = testbed.compute_marching_cubes_mesh(resolution=get_marching_cubes_res(512, testbed.aabb), aabb=testbed.aabb, thresh=2)
```

09:50:35 INFO     #vertices=14423155 #triangles=25967014

RuntimeError                              Traceback (most recent call last)
Cell In[13], line 10
      7 	res3d = round_up_to_base(res3d.astype(np.int32), 16)
      8 	return res3d
---> 10 mc = testbed.compute_marching_cubes_mesh(resolution=get_marching_cubes_res(512, testbed.aabb), aabb=testbed.aabb, thresh=2)

RuntimeError: /home/mehi/Nerfs/instant-ngp/dependencies/tiny-cuda-nn/include/tiny-cuda-nn/cutlass_matmul.h:332 status failed with error Error Internal

**Reply by Fred3D-tech** [reply]
> I tried playing around with this code to compmute the marching cubes mesh for multiple different chunks of the NeRF data by changing the aabb bounds value and whilst it seemed to crop the mesh correctly the result is a bit messed up both with the point positions/resolution and the vertex colours
> 
> do you have any suggestion on how I could export the point cloud for a given region in the testbed bounds by changing the above code?

I have the same question. How to change the testbed bounds without messing everything ? would it be possible to generate a high res point cloud by generating smaller blocks/parts ?
By the way, thanks a lot for the code.

**Reply by kafei123456** [reply]
Excuse me, have you solved the problems? : " generate a high res point cloud"
